### ISSUE TITLE
Can't open CR3 files From Canon EOS R

### ISSUE THREAD

**Issue by B-Souty** [darktable-org/darktable#9559]
Hi everyone,

I'm am unable to import CR3 files from my new Canon EOS R. I saw in the release note of Darktable 3.6.0 that work toward full support of CR3 had begun provided the correct exiv2 library version was found.

I built exiv2 library from source with bmff support and installed Draktable from my package manager but I am still unable to open CR3 files.

I also tried building darktable from source using the `canon_cr3` branch of rawspeed but I kept getting error message so I gave up on that.

**Comment by piratenpanda** [comment]
CR3 files are not yet fully supported in 3.6 and without compiling it yourself or using prepackaged versions with cr3 support it is not expected to work yet

**Comment by B-Souty** [comment]
> CR3 files are not yet fully supported in 3.6 and without compiling it yourself or using prepackaged versions with cr3 support it is not expected to work yet. If you need help compiling ping me on IRC/Matrix

Thanks for offering your help, I won't be able to go on IRC right now, only in a few hours or tomorrow, I'll look around for you.

**Comment by piratenpanda** [comment]
This should do when you already have exiv2 with BMFF support installed:
https://github.com/darktable-org/darktable/issues/2170#issuecomment-854006616

**Comment by B-Souty** [comment]
@piratenpanda This is the comment I was following to do the install initially and I kept getting error during compiling, I was able to fix most of them but got stuck on the last one.

**Comment by piratenpanda** [comment]
Which distro and what's missing?

**Comment by B-Souty** [comment]
@piratenpanda I'm running Ubuntu 20.04 and it seems like pretty much everything is missing... Last time I tried I got stuck on SDL2. I was finally able to install it and now I'm missing llvm-11. I am downloading that as I type but I'm pretty sure on the next run I'll be missing something else. 

I'm not used to build software from source, I usually simply use binaries. I feels like I'm doing something wrong having to install so many different dependencies.

**Comment by B-Souty** [comment]
Yeah I'm lost now, after installing llvm11 it was then complaining about missing llvm12. Which I installed and now the only error message I see when running `cmake ..` is:

```
CMake Error at src/external/CMakeLists.txt:2 (add_subdirectory):
  The source directory

    /home/<redacted>/git_projects/darktable/src/external/libxcf

  does not contain a CMakeLists.txt file.
```

And the only warning about packages not found is:

```
-- The following OPTIONAL packages have not been found:

 * libavif (required version >= 0.8.2)
 * GMIC
 * Lua54 (required version >= 5.4)
```


I check the `src/external/libxcf` and indeed it was empy. I ran git submodule update which repopulated it but running `cmake ..` again resulted in the same error and the `src/external/libxcf` being emptied. 

I also see a bunch of errors in `/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeError.log` but those ones don't make sense to me.

```
Performing C SOURCE FILE Test CMAKE_HAVE_LIBC_PTHREAD failed with the following output:
Change Dir: /home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_ae4b9/fast && /usr/bin/make -f CMakeFiles/cmTC_ae4b9.dir/build.make CMakeFiles/cmTC_ae4b9.dir/build
make[1]: Entering directory '/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_ae4b9.dir/src.c.o
/usr/bin/cc   -DCMAKE_HAVE_LIBC_PTHREAD -fPIE   -o CMakeFiles/cmTC_ae4b9.dir/src.c.o   -c /home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp/src.c
Linking C executable cmTC_ae4b9
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_ae4b9.dir/link.txt --verbose=1
/usr/bin/cc  -DCMAKE_HAVE_LIBC_PTHREAD    CMakeFiles/cmTC_ae4b9.dir/src.c.o  -o cmTC_ae4b9 
/usr/bin/ld: CMakeFiles/cmTC_ae4b9.dir/src.c.o: in function `main':
src.c:(.text+0x46): undefined reference to `pthread_create'
/usr/bin/ld: src.c:(.text+0x52): undefined reference to `pthread_detach'
/usr/bin/ld: src.c:(.text+0x63): undefined reference to `pthread_join'
collect2: error: ld returned 1 exit status
make[1]: *** [CMakeFiles/cmTC_ae4b9.dir/build.make:87: cmTC_ae4b9] Error 1
make[1]: Leaving directory '/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp'
make: *** [Makefile:121: cmTC_ae4b9/fast] Error 2


Source file was:
#include <pthread.h>

void* test_func(void* data)
{
  return data;
}

int main(void)
{
  pthread_t thread;
  pthread_create(&thread, NULL, test_func, NULL);
  pthread_detach(thread);
  pthread_join(thread, NULL);
  pthread_atfork(NULL, NULL, NULL);
  pthread_exit(NULL);

  return 0;
}

Determining if the function pthread_create exists in the pthreads failed with the following output:
Change Dir: /home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_6fda5/fast && /usr/bin/make -f CMakeFiles/cmTC_6fda5.dir/build.make CMakeFiles/cmTC_6fda5.dir/build
make[1]: Entering directory '/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_6fda5.dir/CheckFunctionExists.c.o
/usr/bin/cc   -DCHECK_FUNCTION_EXISTS=pthread_create -fPIE   -o CMakeFiles/cmTC_6fda5.dir/CheckFunctionExists.c.o   -c /usr/share/cmake-3.16/Modules/CheckFunctionExists.c
Linking C executable cmTC_6fda5
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_6fda5.dir/link.txt --verbose=1
/usr/bin/cc  -DCHECK_FUNCTION_EXISTS=pthread_create    CMakeFiles/cmTC_6fda5.dir/CheckFunctionExists.c.o  -o cmTC_6fda5  -lpthreads 
/usr/bin/ld: cannot find -lpthreads
collect2: error: ld returned 1 exit status
make[1]: *** [CMakeFiles/cmTC_6fda5.dir/build.make:87: cmTC_6fda5] Error 1
make[1]: Leaving directory '/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp'
make: *** [Makefile:121: cmTC_6fda5/fast] Error 2



Performing C SOURCE FILE Test HAVE_THREAD_RWLOCK_ARCH_T_NR_READERS failed with the following output:
Change Dir: /home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_d9213/fast && /usr/bin/make -f CMakeFiles/cmTC_d9213.dir/build.make CMakeFiles/cmTC_d9213.dir/build
make[1]: Entering directory '/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_d9213.dir/src.c.o
/usr/bin/cc   -DHAVE_THREAD_RWLOCK_ARCH_T_NR_READERS -fPIE   -o CMakeFiles/cmTC_d9213.dir/src.c.o   -c /home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp/src.c
/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp/src.c: In function ‘main’:
/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp/src.c:6:55: error: ‘struct __pthread_rwlock_arch_t’ has no member named ‘__nr_readers’; did you mean ‘__readers’?
    6 |   (void)sizeof(((struct __pthread_rwlock_arch_t *)0)->__nr_readers);
      |                                                       ^~~~~~~~~~~~
      |                                                       __readers
make[1]: *** [CMakeFiles/cmTC_d9213.dir/build.make:66: CMakeFiles/cmTC_d9213.dir/src.c.o] Error 1
make[1]: Leaving directory '/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp'
make: *** [Makefile:121: cmTC_d9213/fast] Error 2


Source file was:

#include <pthread.h>

int main()
{
  (void)sizeof(((struct __pthread_rwlock_arch_t *)0)->__nr_readers);
  return 0;
}

Performing C SOURCE FILE Test C_COMPILER_UNDERSTANDS_-Wthread-safety failed with the following output:
Change Dir: /home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_2f1c0/fast && /usr/bin/make -f CMakeFiles/cmTC_2f1c0.dir/build.make CMakeFiles/cmTC_2f1c0.dir/build
make[1]: Entering directory '/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_2f1c0.dir/src.c.o
/usr/bin/cc   -Wall -Wformat -Wformat-security -Wshadow -Wtype-limits -Wvla -Wold-style-declaration -DC_COMPILER_UNDERSTANDS_-Wthread-safety  -Wall -Wformat -Wformat-security -Wshadow -Wtype-limits -Wvla -Wold-style-declaration -fPIE   -Wthread-safety -o CMakeFiles/cmTC_2f1c0.dir/src.c.o   -c /home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp/src.c
cc: error: unrecognized command line option '-Wthread-safety'
make[1]: *** [CMakeFiles/cmTC_2f1c0.dir/build.make:66: CMakeFiles/cmTC_2f1c0.dir/src.c.o] Error 1
make[1]: Leaving directory '/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp'
make: *** [Makefile:121: cmTC_2f1c0/fast] Error 2


Source file was:
int main(void) { return 0; }
Performing C++ SOURCE FILE Test CXX_COMPILER_UNDERSTANDS_-Wthread-safety failed with the following output:
Change Dir: /home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_81aac/fast && /usr/bin/make -f CMakeFiles/cmTC_81aac.dir/build.make CMakeFiles/cmTC_81aac.dir/build
make[1]: Entering directory '/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_81aac.dir/src.cxx.o
/usr/bin/c++    -Wall -Wformat -Wformat-security -Wshadow -Wtype-limits -Wvla -DCXX_COMPILER_UNDERSTANDS_-Wthread-safety  -Wall -Wformat -Wformat-security -Wshadow -Wtype-limits -Wvla -fPIE   -Wthread-safety -o CMakeFiles/cmTC_81aac.dir/src.cxx.o -c /home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp/src.cxx
c++: error: unrecognized command line option '-Wthread-safety'
make[1]: *** [CMakeFiles/cmTC_81aac.dir/build.make:66: CMakeFiles/cmTC_81aac.dir/src.cxx.o] Error 1
make[1]: Leaving directory '/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp'
make: *** [Makefile:121: cmTC_81aac/fast] Error 2


Source file was:
int main() { return 0; }
Performing C++ SOURCE FILE Test RAWSPEED_HAVE_CXX_FLAG_WOLD_STYLE_CASTS failed with the following output:
Change Dir: /home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_746be/fast && /usr/bin/make -f CMakeFiles/cmTC_746be.dir/build.make CMakeFiles/cmTC_746be.dir/build
make[1]: Entering directory '/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_746be.dir/src.cxx.o
/usr/bin/c++    -Wall -Wformat -Wformat-security -Wshadow -Wtype-limits -Wvla -Wno-unknown-pragmas -Wno-error=varargs -Wno-format-truncation -Wno-error=address-of-packed-member -w -std=c++14 -Wall -Wextra -Wcast-qual -Wextra -Wextra-semi -Wformat=2 -DRAWSPEED_HAVE_CXX_FLAG_WOLD_STYLE_CASTS  -Wall -Wformat -Wformat-security -Wshadow -Wtype-limits -Wvla -Wno-unknown-pragmas -Wno-error=varargs -Wno-format-truncation -Wno-error=address-of-packed-member -w -std=c++14 -Wall -Wextra -Wcast-qual -Wextra -Wextra-semi -Wformat=2 -fPIE   -Wold-style-casts -std=c++14 -o CMakeFiles/cmTC_746be.dir/src.cxx.o -c /home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp/src.cxx
c++: error: unrecognized command line option '-Wold-style-casts'; did you mean '-Wold-style-cast'?
make[1]: *** [CMakeFiles/cmTC_746be.dir/build.make:66: CMakeFiles/cmTC_746be.dir/src.cxx.o] Error 1
make[1]: Leaving directory '/home/<redacted>/git_projects/darktable/build/CMakeFiles/CMakeTmp'
make: *** [Makefile:121: cmTC_746be/fast] Error 2


Source file was:
int main() { return 0; }
```

**Comment by piratenpanda** [comment]
Did you install all the packages mentioned here? https://github.com/darktable-org/darktable/wiki/Build-instructions-for-Ubuntu

**Comment by B-Souty** [comment]
@piratenpanda It works now. My issue was with `libxcf`. I was too tired yesterday and didn't properly fixed the issue. I thought I did fix the missing files but not really. `git submodule update --init` didn't help. For some reason the files were deleted **and** staged for commit. I had to unstage them and restore using git command, no idea how they ended up in this state.

In any case I was able to compile with simply `cmake -DCMAKE_INSTALL_PREFIX=/opt/darktable_fromsource/ ..`. `make` and `make install` worked without any issue and my darktable is now able to open CR3 file. 

Thanks for all your help.
